# Plotting genomic coordinates with the BioConductor Gviz package
---
*9/8/2016*

These are my notes on how to plot a test dataset generated by Joel Weller using the R bioconductor package ["Gviz".](http://bioconductor.org/packages/release/bioc/html/Gviz.html)

All data files are located in this directory on my workstation:
> /home/dbickhart/share/side_projects/joel_granddaughter_design/data_plots

First, converting data to a usable format:

```bash
perl -lane '$e = $F[0] + 1; print "chr5\t$F[0]\t$e\t$F[1]";' < chr5_region_effects.list > chr5_region_effects.bed

perl -lane 'if($F[0] eq "chr5" && $F[1] < 100115000 && $F[2] > 84223000){print $_;}' < ~/share/umd3_data/gap.bed > chr5_gap_regions.bed
```

OK, so the goal is to display as many of these tracks as possible:
* Ideogram (probably blank!)
* Chr position axis
* Gaps
* Gene track
* Data track (the effect sizes that Joel gave me)

Let's load the data and generate the image.

```R
library(Gviz)

# Identifying the reference database
genomes <- ucscGenomes()
genomes[genomes$species %in% "Cow",]
# ...
# 25 bosTau6     Cow Nov. 2009                University of Maryland v3.1
# ...

# read and organize data
bed <- read.delim("chr5_region_effects.bed", header = FALSE)
bed.granges <- GRanges(seqnames = bed$V1, ranges = IRanges(start = bed$V2, end = bed$V3), score = bed$V4)
chr <- chr5
start <- 84223000
end <- 100115000
gaps <- read.delim("chr5_gap_regions.bed", header=FALSE)
gaps.ir <- GRanges(seqnames = gaps$V1, ranges = IRanges(start = gaps$V2, end = gaps$V3, names = gaps$V4))

# Creating ideogram track
ideo <- IdeogramTrack(genome="bosTau6", chromosome="chr5")

# Biomart gene track
biomart <- BiomartGeneRegionTrack(genome = "bosTau6", chromosome = chr, start = start, end = end, name = "ENSEMBL")

# Axis track
axis <- GenomeAxisTrack(range = gaps.ir)

# Data track
plot <- DataTrack(bed.granges, name = "-Log10 p")

# Now to tie all the tracks together
plotTracks(list(ideo, axis, biomart, plot), from = start, to = end, type = c("l", "g"), stacking = "squish", groupAnnotation = "id", showOverplotting = TRUE)
dev.copy2pdf(file = "chr5_region_log10p_plot.pdf", useDingbats = FALSE)
```

It is relatively easy to tie all of the information into a plot region, but manipulating that plot can take time. I had to sacrifice the gene annotation IDs in order to avoid crowding out the -log10 p data plot.