# Creating accurate scaffolds from BAC-PACBIO sequencing
---
*1/19/2016*

These are my notes on taking existing BAC sequences from the chr18 bac clones, polishing them and assembling them into larger contiguous segments (or at least scaffolding them!).

The quivered BAC clone contigs are unrefined, and likely need pilon correction. I'm going to create a dummy fasta file (reference genome + extra segments to be polished) and Pilon-correct it. Then I'm going to use Velvet to try to do a quick graph-based assembly. If that fails, I'll go pure consensus-overlap (Celera?).

#### Here are the pipeline steps:
1. Identify regions on the original reference genome that are covered by the BAC sequences and remove them
2. Place BAC fasta entries at the end of the reference genome as extra "chrs"
3. Align reads from an individual to the segments
4. Run Pilon on the BAM file (or a subset of the BAM with only the clones we're interested in?)
5. Extract the corrected entries
6. Perform OLC or graph-based assembly

Let's start by getting things ready and proceeding in a step-wise fashion with the existing BAC sequences from John's data.

Files:
* LIB14363_unitig_3_oriented_vector_trimmed.fasta  
* LIB14414_unitig_273.fasta  
* LIB14435_unitig_94_vector_trim.fasta

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs

```bash
# I'm going to test to see if I can identify the regions of the chromosome that need to be removed from a simple alignment
cat *.fasta > pilon_testrun/combined_chr18_fastas.fa
bwa mem ../reference/umd3_kary_unmask_ngap.fa pilon_testrun/combined_chr18_fastas.fa > pilon_testrun/combined_chr18_fastas.sam

perl -lane 'if($F[0] =~ /^@/ || $F[1] > 16){next;}else{print "$F[1]\t$F[2]\t$F[3]\t$F[4]\t" . (length($F[9]) + $F[3]);}' < pilon_testrun/combined_chr18_fastas.sam
	16      chr18   57435857        60      57626394
	0       chr18   57565716        60      57716083
	16      chr18   57676772        60      57835648

# This is obviously not the full picture because the alignments aren't complete, but it's a good start!
# Creating the mask bed file so that we can exclude this region from alignment
echo -e "chr18\t57435857\t57835648" > pilon_testrun/chr18_mask_location.bed
~/bedtools-2.17.0/bin/maskFastaFromBed -fi ../reference/umd3_kary_unmask_ngap.fa -bed pilon_testrun/chr18_mask_location.bed -fo pilon_testrun/umd3_chr18_region_mask.fa

# Now to make and finalize the reference file
cat pilon_testrun/umd3_chr18_region_mask.fa pilon_testrun/combined_chr18_fastas.fa > pilon_testrun/umd3_chr18_masked_combined_unitigs.fa
```

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
bwa index umd3_chr18_masked_combined_unitigs.fa
samtools faidx umd3_chr18_masked_combined_unitigs.fa
perl ~/perl_toolchain/sequence_data_pipeline/runMergedBamPipeline.pl --fastqs ../Arlinda-Chief.10x.spreadsheet.tab --reference umd3_chr18_masked_combined_unitigs.fa --config test_pipeline.cnfg --threads 15 --output arlinda_chr18

# Oops! My config file had an improper ParallelGCThreads argument!
perl ~/perl_toolchain/sequence_data_scripts/bamMergeUtility.pl -d Arlinda-Chief -o arlinda_chr18_merged.bam -n 10

samtools view -H arlinda_chr18/arlinda_chr18_merged.bam | grep '@SQ' | tail
# Supposedly, I can pull space separated chromosomes using samtools
# Let's try

samtools view arlinda_chr18/arlinda_chr18_merged.bam LIB14363_unitig_3 LIB14414_unitig_273 LIB14435_unitig_94 | tail
# Looks like it worked! Generating a bam

samtools view -hb arlinda_chr18/arlinda_chr18_merged.bam LIB14363_unitig_3 LIB14414_unitig_273 LIB14435_unitig_94 > arlinda_chr18/arlinda_chr18_merged_subsectioned.bam
samtools index arlinda_chr18/arlinda_chr18_merged_subsectioned.bam
```

Here's where things get dicey. I don't want Pilon to correct the entire reference (it'll probably take all day!) but I also don't want it to crap out immediately when it checks the BAM header against the fastas that I'm trying to correct. I'm going to demo it with the subset of contigs that I want to correct, and if that fails, I'll try to reformat the bam header in order to trick the program into correcting only the reads that aligned to the specific contigs.

```bash
## Demo 1: subset reference fasta
samtools faidx combined_chr18_fastas.fa

~/jdk1.8.0_05/bin/java -jar ~/bin/pilon-1.13.jar --genome combined_chr18_fastas.fa --frags arlinda_chr18/arlinda_chr18_merged_subsectioned.bam --output combined_chr18_fixed --changes --vcf --diploid

# It worked! And it took only a few minutes!
perl -e '%h; while(<>){chomp; @s = split(/\s+/); if(length($s[2]) == 1 && length($s[3]) == 1 && $s[2] ne "\." && $s[3] ne "\."){$h{"SNP"} += 1;}else{ $h{"INDEL"} += 1;}} print "SNP\t" . $h{"SNP"} . "\n"; print "INDEL\t" . $h{"INDEL"} . "\n"; print "\n";' < combined_chr18_fixed.changes
```

There were several files that were generated by the above commands:
* combined_chr18_fixed.changes  <- This was a log file with all of the changes made
* combined_chr18_fixed.fasta    <- This was the corrected sequence
* combined_chr18_fixed.vcf      <- This is a vcf format of the changes file

#### Data summary

| Dataset | Metric | Value |
| :--- | :--- | ---: |
Original | Length | 499,780 bp
Pilon | Length | 499,711 bp 
Pilon | SNPs | 609
Pilon | INDELs | 115

## Assembling contigs into larger contigs

Now I'm going to take the error-corrected (and uncorrected!) fastas and attempt to merge them into larger contigs. We'll try velvet first, and then Celera if needed. We'll see what works!

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# Note: had to recompile with "make 'LONGSEQUENCES=1'"
~/velvet_1.2.10/velveth pilon_velvet 11 -fasta -long combined_chr18_fixed.fasta
~/velvet_1.2.10/velvetg pilon_velvet

# Damn, it make a ton of contigs, Let's try a larger hash distance
rm pilon_velvet/*
~/velvet_1.2.10/velveth pilon_velvet 31 -fasta -long combined_chr18_fixed.fasta
~/velvet_1.2.10/velvetg pilon_velvet

# Still 1977 nodes. Not worth it.
```

OK, now I'm going to try AMOS (specifically minimus from that pipeline).

```bash
~/amos-3.1.0/bin/toAmos -s combined_chr18_fixed.fasta -o combined_chr18_fixed.afg
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=1
	 The log file is: combined_chr18_fixed.runAmos.log
	Doing step 10:  Building AMOS bank & Dumping reads
	Doing step 11
	Doing step 12
	Doing step 13
	Doing step 20:  Getting overlaps
	Doing step 21
	Command: /usr/local/bin/show-coords -H -c -l -o -r -I 94 combined_chr18_fixed.delta | /home/dbickhart/amos-3.1.0/bin/nucmerAnnotate | egrep 'BEGIN|END|CONTAIN|IDENTITY' > combined_chr18_fixed.coords  exited with status: 1
```

Damn thing needs everything installed to specific directories! I edited the minimus2 executable (it was a shell script) to remove the hardlinks to nucmer entries.

```bash
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=1

# Oops! It turns out that I should rename the original fasta file because it is overwritten by minimus2!
samtools faidx combined_chr18_fixed.fasta
head combined_chr18_fixed.fasta.fai
1       316083  3       60      61
mv combined_chr18_fixed.fasta combined_chr18_fixed.mini_refcount1.fasta

# Hmm, the refcount field may be limiting. I'm setting it to the default to do an all-vs-all alignment
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=0
head combined_chr18_fixed.fasta.fai
1       316083  3       60      61

# Same result!
diff combined_chr18_fixed.fasta combined_chr18_fixed.mini_refcount1.fasta

# No differences!
```

The singular contig represents a 183,628 bp reduction in the naieve concatenated length of the three BAC contigs. 

Let's align this to Chr18 to see how this will work out. I'm interested to see the nucmer lineup and to see if we're hitting the right target region.

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# getting the chr18 fasta ready
samtools faidx ../../reference/umd3_kary_unmask_ngap.fa chr18 > umd3_chr18_unmasked.fa

# This was taken from my notes file: Notes_20150831_dbick_goat_igc_region_annotation_scaffolding.md
nucmer -mumref -l 100 -c 1000 -p chr18_contig_align_test umd3_chr18_unmasked.fa combined_chr18_fixed.fasta
delta-filter -i 95 -o 95 chr18_contig_align_test.delta > chr18_contig_align_test.best.delta
dnadiff -p chr18_contig_align_test.dna -d chr18_contig_align_test.best.delta

mummerplot -p chr18_contig_align_test.mum --large --fat --png -Q combined_chr18_fixed.fasta chr18_contig_align_test.dna.1delta
cat chr18_contig_align_test.mum.gp | awk '{if (match($0, "len=")) { print substr($1, 1, index($1, "_")-1)"\" "$2" "$3} else print $0}'  > chr18_contig_align_test.mum.fixed.gp

# Then I had to change the terminal to emf and comment out mouse settings
gnuplot -geometry 900x900+0+0 -title john_contig chr18_contig_align_test.mum.fixed.gp
```

Hmm, close, but the reference sequence really crowds out the plot here. I need to extract only the reference sequence I need.

```bash
mkdir full_chr_18_fixed
mv chr18_contig_align_test.* full_chr_18_fixed/

samtools faidx ../../reference/umd3_kary_unmask_ngap.fa chr18:57300000-58000000 > umd3_chr18_573_580_unmasked.fa

nucmer -mumref -l 100 -c 1000 -p chr18_subsection_align_test umd3_chr18_573_580_unmasked.fa combined_chr18_fixed.fasta
delta-filter -i 95 -o 95 chr18_subsection_align_test.delta > chr18_subsection_align_test.best.delta
dnadiff -p chr18_subsection_align_test.dna -d chr18_subsection_align_test.best.delta

mummerplot -p chr18_subsection_align_test.mum --large --fat --png -Q combined_chr18_fixed.fasta chr18_subsection_align_test.dna.1delta
cat chr18_subsection_align_test.mum.gp | awk '{if (match($0, "len=")) { print substr($1, 1, index($1, "_")-1)"\" "$2" "$3} else print $0}' > chr18_subsection_align_test.mum.fixed.gp

gnuplot -geometry 900x900+0+0 -title john_contig chr18_subsection_align_test.mum.fixed.gp

mkdir full_chr_18_subsection
mv chr18_subsection_align_test.* full_chr_18_subsection/
```

It looks like John's contig has ~20kb more sequence near the 150kb mark than the reference, and there is one inversion and several translocations! Ah, damn, bad news is that it is the BAC vector! It turns out that Tim did not vector trim Lib14414_unitig_273.fasta, at the least. I found the vector location from NCBI's vecscreen.

## Getting rid of the vector and trying again.

It's a big waste of time, but I've gotta get rid of the vector to be sure that my assembly is correct.


> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# Starting from the beginning
# PILON
~/jdk1.8.0_05/bin/java -jar ~/bin/pilon-1.13.jar --genome combined_chr18_fastas.fa --frags arlinda_chr18/arlinda_chr18_merged_subsectioned.bam --output combined_chr18_pilon --changes --vcf --diploid

# Removing vector from corrected sequence
samtools faidx combined_chr18_pilon.fasta
samtools faidx combined_chr18_pilon.fasta LIB14414_unitig_273_pilon > /mnt/nfs/nfs2/dbickhart/transfer/lib14414_unitig_273.pilon.fa

samtools faidx combined_chr18_pilon.fasta LIB14363_unitig_3_pilon LIB14414_unitig_273_pilon:1-141764 LIB14435_unitig_94_pilon > combined_chr18_pilon.novec.fasta
# This was after NCBI vec screening

# AMOS and Minimus2
~/amos-3.1.0/bin/toAmos -s combined_chr18_pilon.novec.fasta -o combined_chr18_pilon.novec.fixed.afg
mv combined_chr18_pilon.novec.fasta combined_chr18_pilon.novec.fasta.orig
rm combined_chr18_pilon.novec.fasta.fai
~/amos-3.1.0/bin/minimus2 combined_chr18_pilon.novec.fixed -D REFCOUNT=0

samtools faidx combined_chr18_pilon.novec.fixed.fasta
head combined_chr18_pilon.novec.fixed.fasta.fai
1       307495  3       60      61

# Interesting... isn't this the approximate size of the vector?
# I'll do a nucmer plot of this against the previous merged fasta to see how that processed as well

## nucmer against chr18 subsection
nucmer -mumref -l 100 -c 1000 -p chr18_subsection_align_novec umd3_chr18_573_580_unmasked.fa combined_chr18_pilon.novec.fixed.fasta
delta-filter -i 95 -o 95 chr18_subsection_align_novec.delta > chr18_subsection_align_novec.best.delta
dnadiff -p chr18_subsection_align_novec.dna -d chr18_subsection_align_novec.best.delta
mummerplot -p chr18_subsection_align_novec.mum --large --fat --png -Q combined_chr18_pilon.novec.fixed.fasta chr18_subsection_align_novec.dna.1delta
cat chr18_subsection_align_novec.mum.gp | awk '{if (match($0, "len=")) { print substr($1, 1, index($1, "_")-1)"\" "$2" "$3} else print $0}' > chr18_subsection_align_novec.mum.fixed.gp

gnuplot -geometry 900x900+0+0 -title john_contig chr18_subsection_align_novec.mum.fixed.gp
mv chr18_subsection_align_novec.mum.png chr18_subsection_align_novec.mum.emf
```

There's still a section of the genome that doesn't match the UMD3.1 reference. Let's try to run my subsection alignment program on the master, corrected fasta to see what pops out.

```bash
perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f combined_chr18_pilon.novec.fixed.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o combined_chr18_pilon.novec.align.tab

perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f combined_chr18_pilon.novec.fixed.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o combined_chr18_pilon.novec.align.tab
loaded fasta file!
[M::main_mem] read 308 sequences (307494 bp)...
[M::mem_process_seqs] Processed 308 reads in 3.477 CPU sec, 3.392 real sec
[main] Version: 0.7.10-r789
[main] CMD: bwa mem ../../reference/umd3_kary_unmask_ngap.fa temp.fq
[main] Real time: 56.742 sec; CPU: 9.354 sec
Longest aligments:      chr     start   end     length
                        chr1    20951590        153283269       132331679
                        chr2    16512761        125716377       109203616
                        chr11   48330260        107023187       58692927
                        chr5    29133071        85219301        56086230
                        chr16   5146193 55925864        50779671
                        chr23   27321674        52492537        25170863
                        chr13   3668010 27796611        24128601
                        chrX    74404074        96914001        22509927
                        chr18   40333783        62181617        21847834
                        chr17   66030015        73910427        7880412
                        chr3    2051723 2052723 1000
                        chr26   32458140        32459140        1000
                        chr8    12875966        12876966        1000
                        chr29   41546141        41547141        1000
                        chr14   56806974        56807974        1000
                        chr24   13511406        13511520        114
                        chr27   22840110        22840160        50
                        chr28   6264630 6264677 47
                        chr7    25614570        25614615        45
                        chr10   87992202        87992246        44
Output in: combined_chr18_pilon.novec.align.tab
```
The data really just confirmed the nucmer aligns, but with the added bonus of BWA mem mappings of the start and end of the contig to chr18.

## Testing the pipeline on un-corrected fastas 

I'm not happy to just try the pilon data, let's try assembling the uncorrected fastas and plotting that.

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# AMOS
~/amos-3.1.0/bin/toAmos -s combined_chr18_fasta_novec.fa -o combined_chr18_fasta_novec.assemble.afg
~/amos-3.1.0/bin/minimus2 combined_chr18_fasta_novec.assemble -D REFCOUNT=0

samtools faidx combined_chr18_fasta_novec.assemble.fasta
head combined_chr18_fasta_novec.assemble.fasta.fai
1       307535  3       60      61

# So there is a difference in size at least!
# I made an automation script to run the nucmer plot
sh ../run_nucmer_plot_automation_script.sh umd3_chr18_573_580_unmasked.fa combined_chr18_fasta_novec.assemble.fasta

gnuplot -geometry 900x900+0+0 -title john_nopilon_contig combined_chr18_fasta_novec.mum.fixed.gp
```

There were just some minor changes with the sequence towards the 3' end of the assembly, notably the position of a large mum around the 57.65 Mb in the Pilon corrected version that is expanded, and a region in the uncorrected contig that maps towards the end of the contig. I suspect that these are due to poor read alignments after the correction.

<a name="holstein"></a>
## Holstein BACs

OK, the first assembly showed little difference between UMD3 apart from an inversion and some expanded repeats. I'm going to see if the RPCI-42 library (Holstein) has any differences from the Domino BAC sequence.

Here are the files that will be used:
* LIB14397_unitig_1_vector_trimmed.fasta        
* LIB14414_unitig_273.fasta 
* LIB14398_unitig_305_quiver.Vector.Trim.fasta  
* LIB14435_unitig_94_vector_trim.fasta

**NOTE:** the lib14414 unitig is not vector trimmed! Let's do that now.

NCBI's [vecscreen](http://www.ncbi.nlm.nih.gov/tools/vecscreen/) shows that the last 10kb are the vector. Specifically these bases: 141780-150367.

Trimming now...

> pwd: /home/dbickhart/share/grants/immune_gene_cluster_grant/assemblies/chr18_bacs/rpci_42

```bash
samtools faidx LIB14414_unitig_273.fasta
samtools faidx LIB14414_unitig_273.fasta LIB14414_unitig_273:1-141780 | perl -e '<>; print ">LIB14414_unitig_273_trimmed\n"; while(<>){print $_;}' > LIB14414_unitig_273.trimmed.fasta
```

From my notes, 14414 did not assemble well, and may overlap the same region as 14398. Here is the appropriate table:

Library |	Region Name	| Clone Name	| End Coords	| Notes	| Library	| Results
:--- | :--- | :--- | :--- | :--- | :--- | :---
RPCI-42	|chr18|	RPCI42_154M1|	chr18:57,371,161-57,531,510|	Flanking 5' region|	14397|	finished
RPCI-42	|chr18|	RPCI42_118F24|	chr18:57,548,063-57,704,285|	Overlaps SNP|	14414|	Did not assemble well
RPCI-42	|chr18|	RPCI42_118B22|	chr18:57,548,064-57,704,269|	Perhaps a clone of the previous region?|	14398|	finished
RPCI-42	|chr18|	RPCI42_3D15|	chr18:57,657,380-57,806,510|	Flanking 3' region|	14435|	finished

My thoughts: three assemblies might be best here. Assemble with 14414, with 14398 and with both.  

Lets start with the entirety of the pipeline and get Pilon-corrected assemblies before Amos overlap.

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/holstein_bacs

```bash
# Removing target chr18 regions
cat ../LIB14397_unitig_1_vector_trimmed.fasta ../LIB14398_unitig_305_quiver.Vector.Trim.fasta ../LIB14414_unitig_273.vectortrim.fasta ../LIB14435_unitig_94_vector_trim.fasta > rpci42_bac.fasta
bwa mem ../../reference/umd3_kary_unmask_ngap.fa rpci42_bac.fasta > rpci42_bacs_uncorrected_align.sam

perl -lane 'if($F[0] =~ /^@/ || $F[1] > 16){next;}else{print "$F[0]\t$F[1]\t$F[2]\t$F[3]\t$F[4]\t" . (length($F[9]) + $F[3]);}' < rpci42_bacs_uncorrected_align.sam
	LIB14397_unitig_1        0       chr18   57381124        60      57541326
	LIB14398_unitig_305     16      chr18   57565716        60      57729084
	LIB14414_unitig_273     0       chr18   57565716        60      57707495
	LIB14435_unitig_94     16      chr18   57676772        60      57835648

# Looks like the region is chr18:57381124-57835648
# NOTE: the two "clones" that Tim mentioned differ in end sequence alignments by ~ 20kb
# Masking the fasta
echo -e "chr18\t57381124\t57835648" > rpci42_bac_mask.bed
~/bedtools-2.17.0/bin/maskFastaFromBed -fi ../../reference/umd3_kary_unmask_ngap.fa -bed rpci42_bac_mask.bed -fo chr18_initial_masked.fa

# Making the pilon fasta
cat chr18_initial_masked.fa rpci42_bac.fasta > chr18_combined_unitigs_masked.fa
bwa index chr18_combined_unitigs_masked.fa
samtools faidx chr18_combined_unitigs_masked.fa

# Alignment
perl ~/perl_toolchain/sequence_data_pipeline/runMergedBamPipeline.pl --fastqs ../Arlinda-Chief.10x.spreadsheet.tab --reference chr18_combined_unitigs_masked.fa --config ../pilon_testrun/test_pipeline.cnfg --threads 15 --output arlinda_rpci42

# Extracting contigs
samtools view -hb arlinda_rpci42/Arlinda-Chief/Arlinda-Chief.merged.bam 'LIB14397_unitig_1_vector_trimmed' 'LIB14398_unitig_305|quiver_Vector_Trim' LIB14414_unitig_273 LIB14435_unitig_94_vector_trim > arlinda_rpci42/Arlinda-Chief.unitigs.bam

# Pilon correction
samtools index /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/holstein_bacs/arlinda_rpci42/Arlinda-Chief.unitigs.bam
~/jdk1.8.0_05/bin/java -jar ~/bin/pilon-1.13.jar --genome chr18_combined_unitigs_masked.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.bam --output chr18_rpci42_unitigs --changes --vcf --diploid

# Pilon crashed -- I suspect its due to the presence of a pipe in the fasta names
perl -ne 'if($_ =~ /^>/){ $_ =~ s/\|/_/g; print $_;}else{print $_;}' < chr18_combined_unitigs_masked.fa > chr18_combined_unitigs_masked.nopipe.fa
perl -ne 'if($_ =~ /^>/){ $_ =~ s/\|/_/g; print $_;}else{print $_;}' < rpci42_bac.fasta > rpci42_bac.corrected.fasta

samtools view -h arlinda_rpci42/Arlinda-Chief.unitigs.bam | perl -ne '$_ =~ s/\|/_/g; print $_;' | samtools view -bS - > arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam
samtools index arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam

~/jdk1.8.0_05/bin/java -Xmx75g -jar ~/bin/pilon-1.13.jar --genome chr18_combined_unitigs_masked.nopipe.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam --output chr18_rpci42_unitigs --changes --vcf --diploid
# OK... still crashing. Let's try this without the pilon correction

# Amos assembly of non-polished contigs
cp rpci42_bac.corrected.fasta rpci42_bac.corrected.fasta.bak
~/amos-3.1.0/bin/toAmos -s rpci42_bac.corrected.fasta -o rpci42_bac.corrected.afg
~/amos-3.1.0/bin/minimus2 rpci42_bac.corrected -D REFCOUNT=0

head rpci42_bac.corrected.fasta.fai
	1       163390  3       60      61

# That's... really weird. Let's see how the alignments turn out
perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f rpci42_bac.corrected.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o rpci42_bac.corrected.fasta.align.tab
# Nope! Doesn't look good.

# Let's try individual pilon targets
samtools faidx rpci42_bac.fasta
# Lib14397_1 worked
~/jdk1.8.0_05/bin/java -Xmx70g -jar ~/bin/pilon-1.16.jar --genome chr18_combined_unitigs_masked.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam --output lib14397_1_pilon --changes --vcf --diploid --targets LIB14397_unitig_1_vector_trimmed

# Lib14398_305 failed
~/jdk1.8.0_05/bin/java -Xmx70g -jar ~/bin/pilon-1.16.jar --genome chr18_combined_unitigs_masked.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam --output lib14398_305_pilon --changes --vcf --diploid --targets LIB14398_unitig_305_quiver_Vector_Trim

# Lib14414_273 worked
~/jdk1.8.0_05/bin/java -Xmx70g -jar ~/bin/pilon-1.16.jar --genome chr18_combined_unitigs_masked.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam --output lib14414_273_pilon --changes --vcf --diploid --targets LIB14414_unitig_273

# Lib14435_94 worked
~/jdk1.8.0_05/bin/java -Xmx70g -jar ~/bin/pilon-1.16.jar --genome chr18_combined_unitigs_masked.fa --frags arlinda_rpci42/Arlinda-Chief.unitigs.corrected.bam --output lib14435_94_pilon --changes --vcf --diploid --targets LIB14435_unitig_94_vector_trim

# Taking the error corrected fasta entries for further analysis
cat lib14397_1_pilon.fasta lib14414_273_pilon.fasta lib14435_94_pilon.fasta > pilon_corrected_rpci14_bac.fa
cp pilon_corrected_rpci14_bac.fa pilon_corrected_rpci14_bac.fa.bak

# Amos assembly of polished contigs
~/amos-3.1.0/bin/toAmos -s pilon_corrected_rpci14_bac.fa -o pilon_corrected_rpci14_bac.afg
~/amos-3.1.0/bin/minimus2 pilon_corrected_rpci14_bac -D REFCOUNT=0
samtools faidx pilon_corrected_rpci14_bac.fasta

# I didn't get a contig, trying to figure out why
```

OK! The problem is in the nucmer coords file, which shows no overlap among the three contigs. Also, in my prior assembly, the third contig didn't assemble into the final product at all! The information is in the "singletons" file. 

First assembly singletons:
* LIB14435_unitig_94_vector_trim

Second assembly singletons:
* LIB14397_unitig_1_vector_trimmed_pilon
* LIB14414_unitig_273_pilon
* LIB14435_unitig_94_vector_trim_pilon

Let's try an assembly alignment to try to sort this out.

```bash
# Lib 14397
perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f lib14397_1_pilon.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o lib14397_1_pilon.fasta.align.tab
Longest aligments:      chr     start   end     length
                        chr18   57371117        57532624        161507
                        *       0       1000    1000
                        chr27   22286043        22286293        250
                        chrX    44952057        44952154        97
                        chr6    113122076       113122147       71
                        chr4    3045190 3045260 70

# Lib 14414
perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f lib14414_273_pilon.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o lib14414_273_pilon.fasta.align.tab
Longest aligments:      chr     start   end     length
                        chr1    20951590        153283269       132331679
                        chr24   13511406        55366739        41855333
                        chrX    101793354       142730599       40937245
                        chr16   5146193 36768842        31622649
                        chr18   40333924        62955483        22621559
                        chr23   27321674        42251667        14929993
                        chr13   3668010 3669010 1000
                        chr15   25930890        25931890        1000
                        chr11   6304540 6305540 1000
                        chr26   32458140        32459140        1000
                        chr6    48897110        48898110        1000
                        chr2    16512761        16513761        1000
						... not so good!
# Lib 14435
perl ~/perl_toolchain/assembly_scripts/alignUnitigSectionsToRef.pl -f lib14435_94_pilon.fasta -r ../../reference/umd3_kary_unmask_ngap.fa -o lib14435_94_pilon.fasta.align.tab
Longest aligments:      chr     start   end     length
                        chr1    20951590        121426804       100475214
                        chr10   15174811        87992246        72817435
                        chrX    74404074        134697683       60293609
                        chr16   5146193 55925864        50779671
                        chr18   15236467        62961880        47725413
                        chr5    22425571        22426571        1000
                        chr26   32458140        32459140        1000
                        chr20   7720770 7721770 1000
                        chr13   78628163        78629163        1000
                        chr4    69969728        69970728        1000
                        chr11   74810900        74811900        1000
                        chr6    67547644        67547816        172
                        chr24   13511406        13511551        145
                        chr23   27321674        27321789        115
                        chr19   53681295        53681355        60
```

OK, the problem is that 14414 has issues and so does 14435. I suspect that they were not properly assembled.
