# Creating accurate scaffolds from BAC-PACBIO sequencing
---
*1/19/2016*

These are my notes on taking existing BAC sequences from the chr18 bac clones, polishing them and assembling them into larger contiguous segments (or at least scaffolding them!).

The quivered BAC clone contigs are unrefined, and likely need pilon correction. I'm going to create a dummy fasta file (reference genome + extra segments to be polished) and Pilon-correct it. Then I'm going to use Velvet to try to do a quick graph-based assembly. If that fails, I'll go pure consensus-overlap (Celera?).

#### Here are the pipeline steps:
1. Identify regions on the original reference genome that are covered by the BAC sequences and remove them
2. Place BAC fasta entries at the end of the reference genome as extra "chrs"
3. Align reads from an individual to the segments
4. Run Pilon on the BAM file (or a subset of the BAM with only the clones we're interested in?)
5. Extract the corrected entries
6. Perform OLC or graph-based assembly

Let's start by getting things ready and proceeding in a step-wise fashion with the existing BAC sequences from John's data.

Files:
* LIB14363_unitig_3_oriented_vector_trimmed.fasta  
* LIB14414_unitig_273.fasta  
* LIB14435_unitig_94_vector_trim.fasta

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs

```bash
# I'm going to test to see if I can identify the regions of the chromosome that need to be removed from a simple alignment
cat *.fasta > pilon_testrun/combined_chr18_fastas.fa
bwa mem ../reference/umd3_kary_unmask_ngap.fa pilon_testrun/combined_chr18_fastas.fa > pilon_testrun/combined_chr18_fastas.sam

perl -lane 'if($F[0] =~ /^@/ || $F[1] > 16){next;}else{print "$F[1]\t$F[2]\t$F[3]\t$F[4]\t" . (length($F[9]) + $F[3]);}' < pilon_testrun/combined_chr18_fastas.sam
	16      chr18   57435857        60      57626394
	0       chr18   57565716        60      57716083
	16      chr18   57676772        60      57835648

# This is obviously not the full picture because the alignments aren't complete, but it's a good start!
# Creating the mask bed file so that we can exclude this region from alignment
echo -e "chr18\t57435857\t57835648" > pilon_testrun/chr18_mask_location.bed
~/bedtools-2.17.0/bin/maskFastaFromBed -fi ../reference/umd3_kary_unmask_ngap.fa -bed pilon_testrun/chr18_mask_location.bed -fo pilon_testrun/umd3_chr18_region_mask.fa

# Now to make and finalize the reference file
cat pilon_testrun/umd3_chr18_region_mask.fa pilon_testrun/combined_chr18_fastas.fa > pilon_testrun/umd3_chr18_masked_combined_unitigs.fa
```

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
bwa index umd3_chr18_masked_combined_unitigs.fa
samtools faidx umd3_chr18_masked_combined_unitigs.fa
perl ~/perl_toolchain/sequence_data_pipeline/runMergedBamPipeline.pl --fastqs ../Arlinda-Chief.10x.spreadsheet.tab --reference umd3_chr18_masked_combined_unitigs.fa --config test_pipeline.cnfg --threads 15 --output arlinda_chr18

# Oops! My config file had an improper ParallelGCThreads argument!
perl ~/perl_toolchain/sequence_data_scripts/bamMergeUtility.pl -d Arlinda-Chief -o arlinda_chr18_merged.bam -n 10

samtools view -H arlinda_chr18/arlinda_chr18_merged.bam | grep '@SQ' | tail
# Supposedly, I can pull space separated chromosomes using samtools
# Let's try

samtools view arlinda_chr18/arlinda_chr18_merged.bam LIB14363_unitig_3 LIB14414_unitig_273 LIB14435_unitig_94 | tail
# Looks like it worked! Generating a bam

samtools view -hb arlinda_chr18/arlinda_chr18_merged.bam LIB14363_unitig_3 LIB14414_unitig_273 LIB14435_unitig_94 > arlinda_chr18/arlinda_chr18_merged_subsectioned.bam
samtools index arlinda_chr18/arlinda_chr18_merged_subsectioned.bam
```

Here's where things get dicy. I don't want Pilon to correct the entire reference (it'll probably take all day!) but I also don't want it to crap out immediately when it checks the BAM header against the fastas that I'm trying to correct. I'm going to demo it with the subset of contigs that I want to correct, and if that fails, I'll try to reformat the bam header in order to trick the program into correcting only the reads that aligned to the specific contigs.

```bash
## Demo 1: subset reference fasta
samtools faidx combined_chr18_fastas.fa

~/jdk1.8.0_05/bin/java -jar ~/bin/pilon-1.13.jar --genome combined_chr18_fastas.fa --frags arlinda_chr18/arlinda_chr18_merged_subsectioned.bam --output combined_chr18_fixed --changes --vcf --diploid

# It worked! And it took only a few minutes!
perl -e '%h; while(<>){chomp; @s = split(/\s+/); if(length($s[2]) == 1 && length($s[3]) == 1 && $s[2] ne "\." && $s[3] ne "\."){$h{"SNP"} += 1;}else{ $h{"INDEL"} += 1;}} print "SNP\t" . $h{"SNP"} . "\n"; print "INDEL\t" . $h{"INDEL"} . "\n"; print "\n";' < combined_chr18_fixed.changes
```

There were several files that were generated by the above commands:
* combined_chr18_fixed.changes  <- This was a log file with all of the changes made
* combined_chr18_fixed.fasta    <- This was the corrected sequence
* combined_chr18_fixed.vcf      <- This is a vcf format of the changes file

#### Data summary

| Dataset | Metric | Value |
| :--- | :--- | ---: |
Original | Length | 499,780 bp
Pilon | Length | 499,711 bp 
Pilon | SNPs | 609
Pilon | INDELs | 115

## Assembling contigs into larger contigs

Now I'm going to take the error-corrected (and uncorrected!) fastas and attempt to merge them into larger contigs. We'll try velvet first, and then Celera if needed. We'll see what works!

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# Note: had to recompile with "make 'LONGSEQUENCES=1'"
~/velvet_1.2.10/velveth pilon_velvet 11 -fasta -long combined_chr18_fixed.fasta
~/velvet_1.2.10/velvetg pilon_velvet

# Damn, it make a ton of contigs, Let's try a larger hash distance
rm pilon_velvet/*
~/velvet_1.2.10/velveth pilon_velvet 31 -fasta -long combined_chr18_fixed.fasta
~/velvet_1.2.10/velvetg pilon_velvet

# Still 1977 nodes. Not worth it.
```

OK, now I'm going to try AMOS (specifically minimus from that pipeline).

```bash
~/amos-3.1.0/bin/toAmos -s combined_chr18_fixed.fasta -o combined_chr18_fixed.afg
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=1
	 The log file is: combined_chr18_fixed.runAmos.log
	Doing step 10:  Building AMOS bank & Dumping reads
	Doing step 11
	Doing step 12
	Doing step 13
	Doing step 20:  Getting overlaps
	Doing step 21
	Command: /usr/local/bin/show-coords -H -c -l -o -r -I 94 combined_chr18_fixed.delta | /home/dbickhart/amos-3.1.0/bin/nucmerAnnotate | egrep 'BEGIN|END|CONTAIN|IDENTITY' > combined_chr18_fixed.coords  exited with status: 1
```

Damn thing needs everything installed to specific directories! I edited the minimus2 executable (it was a shell script) to remove the hardlinks to nucmer entries.

```bash
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=1

# Oops! It turns out that I should rename the original fasta file because it is overwritten by minimus2!
samtools faidx combined_chr18_fixed.fasta
head combined_chr18_fixed.fasta.fai
1       316083  3       60      61
mv combined_chr18_fixed.fasta combined_chr18_fixed.mini_refcount1.fasta

# Hmm, the refcount field may be limiting. I'm setting it to the default to do an all-vs-all alignment
~/amos-3.1.0/bin/minimus2 combined_chr18_fixed -D REFCOUNT=0
head combined_chr18_fixed.fasta.fai
1       316083  3       60      61

# Same result!
diff combined_chr18_fixed.fasta combined_chr18_fixed.mini_refcount1.fasta

# No differences!
```

The singular contig represents a 183,628 bp reduction in the naieve concatenated length of the three BAC contigs. 

Let's align this to Chr18 to see how this will work out. I'm interested to see the nucmer lineup and to see if we're hitting the right target region.

> Blade14: /mnt/iscsi/vnx_gliu_7/john_assembled_contigs/pilon_testrun

```bash
# getting the chr18 fasta ready
samtools faidx ../../reference/umd3_kary_unmask_ngap.fa chr18 > umd3_chr18_unmasked.fa

# This was taken from my notes file: Notes_20150831_dbick_goat_igc_region_annotation_scaffolding.md
nucmer -mumref -l 100 -c 1000 -p chr18_contig_align_test umd3_chr18_unmasked.fa combined_chr18_fixed.fasta
delta-filter -i 95 -o 95 chr18_contig_align_test.delta > chr18_contig_align_test.best.delta
dnadiff -p chr18_contig_align_test.dna -d chr18_contig_align_test.best.delta

mummerplot -p chr18_contig_align_test.mum --large --fat --png -Q combined_chr18_fixed.fasta chr18_contig_align_test.dna.1delta
cat chr18_contig_align_test.mum.gp | awk '{if (match($0, "len=")) { print substr($1, 1, index($1, "_")-1)"\" "$2" "$3} else print $0}'  > chr18_contig_align_test.mum.fixed.gp

# Then I had to change the terminal to emf and comment out mouse settings
gnuplot -geometry 900x900+0+0 -title john_contig chr18_contig_align_test.mum.fixed.gp
```

Hmm, close, but the reference sequence really crowds out the plot here. I need to extract only the reference sequence I need.

```bash
mkdir full_chr_18_fixed
mv chr18_contig_align_test.* full_chr_18_fixed/

samtools faidx ../../reference/umd3_kary_unmask_ngap.fa chr18:57300000-58000000 > umd3_chr18_573_580_unmasked.fa

nucmer -mumref -l 100 -c 1000 -p chr18_subsection_align_test umd3_chr18_573_580_unmasked.fa combined_chr18_fixed.fasta
delta-filter -i 95 -o 95 chr18_subsection_align_test.delta > chr18_subsection_align_test.best.delta
dnadiff -p chr18_subsection_align_test.dna -d chr18_subsection_align_test.best.delta

mummerplot -p chr18_subsection_align_test.mum --large --fat --png -Q combined_chr18_fixed.fasta chr18_subsection_align_test.dna.1delta
cat chr18_subsection_align_test.mum.gp | awk '{if (match($0, "len=")) { print substr($1, 1, index($1, "_")-1)"\" "$2" "$3} else print $0}' > chr18_subsection_align_test.mum.fixed.gp

gnuplot -geometry 900x900+0+0 -title john_contig chr18_subsection_align_test.mum.fixed.gp
```

It looks like John's contig has ~20kb more sequence near the 150kb mark than the reference, and there is one inversion and several translocations!